packages/protocol/contracts/stability/StableToken.sol

```diff
@@ -527,4 +530,48 @@ contract StableToken is
     return true;
   }
 
+  mapping(address => uint256) reserved;
```

Comment:
I'm worried about the gas implications of the extra writes that using this mapping requires. What do you think about the following implementation?

```
  function debitGas(address from, uint256 value)
    external
    onlyVm
    updateInflationFactor
    onlyWhenNotFrozen
  {
    uint256 units = _valueToUnits(inflationState.factor, value);
    totalSupply_ = totalSupply_.sub(units);
    balances[from] = balances[from].sub(units);
  }

function payGas(address from, address feeRecipient, address gatewayFeeRecipient, uint256 gas, uint256 gasPrice, uint256 gasUsed, uint256 gatewayFee) external onlyVm {
 uint256 gasValue = gas.mul(gasPrice);
  uint256 units = _valueToUnits(gasValue.add(gatewayFee));
  uint256 unitsUsed = _valueToUnits(gasValue);
  uint256 unitsGateway = _valueToUnits(gatewayFee);
  totalSupply_ = totalSupply_.add(units);
  balances[feeRecipient] = balances[feeRecipient].add(unitsUsed);
  balances[gatewayFeeRecipient] = balances[gatewayFeeRecipient].add(unitsGateway);
  balances[from] = balances[from].add(units.sub(unitsUsed).sub(unitsGateway));
  emit Transfer(from, feeRecipient, gasValue);
  emit Transfer(from, gatewayFeeRecipient, gatewayFee);
}
```

Reply:
I guess I should say, the added cost of the additional writes to the storage trie, costs that of course wouldn't directly be passed down to the user since the gas cost to the sum of these operations is hardcoded in BlockChainParameters.

To be fair, the storage trie changes for `reserved` would never actually be written to disk, since any changes are reverted by the end of tx processing.

Reply:
cc @kevjue to get your thoughts here

Reply:
Is there a corresponding PR in the geth code that uses these new functions?  It would help to get more context.

Reply:
It's here: https://github.com/celo-org/celo-blockchain/pull/882

Reply:
I like @asaj 's proposal (perhaps rename `debitGas` to `debitGasFee` and `payGas` to `payGasFee`.  That proposal seems to be much more explicit for the purpose of what's intended as opposed to @mrsmkl 's.

One thing I do like about @mrsmkl 's implementation is that it's explicit about the debited gas.  E.g. it was initially confusing to me in @asaj 's when the totalSupply was decreased in `debitGas`.  So if you go w/ @asaj 's then I recommend adding comments about that.

Reply:
A few comments...

1. We need to `reserve` token for gas. So changing `totalSupply` doesn't make sense.
2. The protocol is very strict that for every `reserve` there's a `credit`. If we use that information we don't actually need to save anything on the contract. At most, you could save a `reservedValue` but since that information can be given to you at the `credit` call. It makes no sense to write to the trie storage.
3. Even if writes to trie storage won't be persisted, it's surely cheaper (computationally wise) to let the node (and not the contract carry the state)

Reply:
So... I'd go with @asaj approach...

I'm a bit uncertain about the names... `debit` and `pay` sound like synonims to me. I would change `debit` for something like `reserveForGas` `lockForGas`... but not totally sure. And maybe chage the `pay` to something like `applyGasPayment`

Also, i would change the signature of the second function to

`applyGasPayment(addresses[3], amounts[3])` where addresses[0] does not emit a transfer since it's the `from`
or 
`applyGasPayment(address from, address feeRecipient, address gatewayFee, uint32 amountFrom, amountFeeR, amountGat..)`

basically, i wouldn't make the contract make the gas price computation

Reply:
> We need to reserve token for gas. So changing totalSupply doesn't make sense.
This was done originally to account for rounding errors. If we have an approach that is resilient to this by calculating the exact same total of units on both ends of the tx, changing total supply won't be necessary. I think this is another argument for the proposal above.

RE changing the second function signature, +1 to either of @mcortesi 's suggestions.

Finally, on naming, suggest `debitGasFee` and `creditGasFees`, which at the very least is consistent with our current naming scheme.

Reply:
Can somebody check that `units.sub(unitsUsed).sub(unitsGateway)` won't fail because fo rounding errors?

Reply:
it won't

Reply:
Just figured out that we need the `totalSupply` anyway.

If the tx call `setInflactionFactor()` then `valueToUnits()` result will change; since inflation factor might have changed.